### PR TITLE
WIP: add yarn specific tasks to gradle build config

### LINKED ISSUE #5177: Yarn not found in generated Jenkins pipeline file
##### **Overview of the issue**
I have used the latest JHipster 4 generator to create a simple (monolithic) app, selecting angular 2 and gradle as build system. Although JHipster is really awesome I get the following error when trying to run the generated build pipeline:
```
:npmInstall
> hippo-4@0.0.0 postinstall /var/jenkins_home/workspace/test-pipeline 
> yarn run webpack:build
sh: 1: yarn: not found
```

##### **Motivation for or Use Case**
I am running jenkins 2.32, if a Jenkinsfile is generated it should work out of the box.

##### **Reproduce the error**
To reproduce you should use latests release JHipster version, create a sample project using **gradle** as the build tool and **angular2** lib (all other options can be default). 
Now try to use the generated Jenkinsfile on a jenkins server where no yarn is installed, the build should fail.

##### **JHipster Version(s)**
4.0.2

##### **JHipster configuration**
```yaml
{
  "generator-jhipster": {
    "jhipsterVersion": "4.0.2",
    "baseName": "hippo4",
    "packageName": "de.joergherbst.hippo",
    "packageFolder": "de/joergherbst/hippo",
    "serverPort": "8080",
    "authenticationType": "session",
    "hibernateCache": "ehcache",
    "clusteredHttpSession": false,
    "websocket": false,
    "databaseType": "sql",
    "devDatabaseType": "h2Disk",
    "prodDatabaseType": "postgresql",
    "searchEngine": false,
    "messageBroker": false,
    "buildTool": "gradle",
    "enableSocialSignIn": false,
    "rememberMeKey": "929ff6b80a3a4d1dbc6da49c965aeda3bd7ed2eb",
    "useSass": false,
    "clientPackageManager": "yarn",
    "applicationType": "monolith",
    "clientFramework": "angular2",
    "testFrameworks": [
      "gatling"
    ],
    "jhiPrefix": "jhi",
    "enableTranslation": true,
    "nativeLanguage": "en",
    "languages": [
      "en",
      "de"
    ]
  }
}
```
##### **Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**
no entity configured

##### **Browsers and Operating System**

Cause this is not an UI issue there is no browser dependency.

### PR DESCRIPTION
**Pull request by atomfrede**
I did not put yarn install into a sperate stage (like `install tools` in maven). If desired we can do that of course.

close #5177

### CODE REVIEW

**Comment by jdubois** [comment]
This won't work on Windows, but I guess we don't care about running a CI server on Windows ?

**Comment by pascalgrimaud** [comment]
@jdubois : no it won't work on Windows, but the initial idea was to provide a sample of Jenkinsfile for Linux. If the user want to run his CI server on Windows, he can replace the command.
Maybe we should document it

@atomfrede : I don't think it will close #5177 as there are some modifications to do for maven too (replace gulp when it's a ng2 project)

**Comment by atomfrede** [comment]
@pascalgrimaud Ah yes, after looking at it a second time you are right. Yarn can be used with ng1, too. Will take a look at it.

File: generators/server/templates/_Jenkinsfile [hunk 1]
```diff
@@ -65,12 +65,16 @@ node {
     stage('clean') {
         sh "./gradlew clean"
     }
-
     <%_ if (applicationType !== 'microservice') { _%>
+        <%_ if (clientPackageManager === 'yarn') { _%>
+    stage('yarn install') {
+        sh "./gradlew yarn_install -PnodeInstall"
```
Comment (kaidohallik) [inline generators/server/templates/_Jenkinsfile]: I tested this with following configuration:
* Jenkins 2.45
* Docker 1.12.6
* Centos 7.3

It throwed error:
```
[4/4] Building fresh packages...
error /var/jenkins_home/workspace/jhipster-jenkins-test/node_modules/yo: Command failed.
Exit code: 1
Command: sh
Arguments: -c yodoctor
Directory: /var/jenkins_home/workspace/jhipster-jenkins-test/node_modules/yo
Output:
Yeoman Doctor
Running sanity checks on your system

✔ Global configuration file is valid
✔ NODE_PATH matches the npm root
✔ Node.js version
✔ No .bowerrc file in home directory
✔ No .yo-rc.json file in home directory
/var/jenkins_home/workspace/jhipster-jenkins-test/node_modules/semver-truncate/index.js:12
		throw new Error('Version ' + version + ' is not valid semver');
		^

Error: Version null is not valid semver
    at module.exports (/var/jenkins_home/workspace/jhipster-jenkins-test/node_modules/semver-truncate/index.js:12:9)
    at /var/jenkins_home/workspace/jhipster-jenkins-test/node_modules/bin-version-check/index.js:20:25
    at /var/jenkins_home/workspace/jhipster-jenkins-test/node_modules/bin-version/index.js:15:3
    at ChildProcess.exithandler (child_process.js:197:7)
    at emitTwo (events.js:106:13)
    at ChildProcess.emit (events.js:191:7)
    at maybeClose (internal/child_process.js:877:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:226:5)
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
:yarn_install FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':yarn_install'.
> Process 'command '/var/jenkins_home/workspace/jhipster-jenkins-test/.gradle/nodejs/node-v6.9.5-linux-x64/bin/node'' finished with non-zero exit value 1

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED
```

My .yo-rc.json:
```
{
  "generator-jhipster": {
    "jhipsterVersion": "4.0.2",
    "baseName": "jhipster",
    "packageName": "com.mycompany.myapp",
    "packageFolder": "com/mycompany/myapp",
    "serverPort": "8080",
    "authenticationType": "session",
    "hibernateCache": "ehcache",
    "clusteredHttpSession": false,
    "websocket": false,
    "databaseType": "sql",
    "devDatabaseType": "h2Disk",
    "prodDatabaseType": "mysql",
    "searchEngine": false,
    "messageBroker": false,
    "serviceDiscoveryType": false,
    "buildTool": "gradle",
    "enableSocialSignIn": false,
    "rememberMeKey": "be20d9ae8447ba5287bdc47245fcd9c8f727ebea",
    "useSass": false,
    "clientPackageManager": "yarn",
    "applicationType": "monolith",
    "clientFramework": "angular2",
    "testFrameworks": [],
    "jhiPrefix": "jhi",
    "enableTranslation": true,
    "nativeLanguage": "en",
    "languages": [
      "en"
    ]
  }
}
```

Yesterday I tried to get work yarn_install but with no success. If appears that it's not possible to get Jenkins pipeline work with yarn_install then in #5177 is changelog for npmInstall (unfortunately mix of yarn and npm) what I finally got work.

**Comment by pascalgrimaud** [comment]
@atomfrede : I don't know if you has worked on this after my comment, but currently I'm working on it, to fix https://github.com/jhipster/generator-jhipster/issues/5188

And I think this Jenkinsfile should be move to ci-cd sub generator

**Comment by cbornet** [comment]
I  remember having an issue with the yarn_install task when installing yo (dependency of generator-jhipster) because it didn't find npm (when you don't have it installed on your CI)

**Comment by cbornet** [comment]
Actually it's the issue @kaidohallik is seing : yo-doctor checks the npm version, gets nothing and semver fails

**Comment by pascalgrimaud** [comment]
I'm closing this, as it is all done here: https://github.com/jhipster/generator-jhipster/pull/5225
Hope you're ok @atomfrede

**Pull request closed without merging**